### DISCUSSION TITLE
Verbräuche mit/ohne PV-Anhebung für WP erfassen

### DISCUSSION THREAD

**Discussion by realmuemmel** [camueller/SmartApplianceEnabler#216] [Installation und Betrieb]
Hallo,

ich nutze aktuell SAE auf meinem Raspberry zusammen mit dem SMA HM2.0, bisher funktioniert alles einwandfrei, danke dafür! Jetzt möchte ich vom SHM die PV-Anhebung meiner Wärmepumpe ansteuern, sobald genug PV-Überschuss vorhanden ist. Den Verbrauch der WP übermittle ich per ModBus von einem eigenen Zähler (SDM72DM-V2 MID) an SAE. Die PV-Anhebung wird per potenzialfreiem Kontakt über ein GPIO-Relais am Raspberry gesteuert. Jetzt habe ich aber folgendes Problem: Die Verbrauchswerte im Sunnyportal werden korrekt ausgelesen und im "Aktuellen Verbrauch" angezeigt, allerdings tauchen diese nie in der Tagesübersicht als Gesamtverbrauch auf. Ich vermute, dass die Verbräuche dort nur auftauchen, wenn der Verbraucher als eingeschaltet markiert ist, was bei mir ja nur in den Phasen der PV-Anhebung der Fall wäre. Die WP kann aber antürlich auch unabhängig von der PV-Anhebung laufen. Diese Verbräuche möchte ich natürlich auch erfassen und im Idealfall für die Schaltentscheidung (Maximalverbauch) mit einfließen lassen.
Gibt es für dieses Problem eine passende Lösung oder wir wird PV-Anhebung üblicherweise in SAE konfiguriert?

Danke und viele Grüße,
Nico

**Comment by premultiply** [comment]
Schau mal hier: https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/MasterSlaveMeter_DE.md
Das sollte nun mit "Master/Slave-Zähler" ohne "Hardwaretricks" möglich sein.

**Comment by realmuemmel** [comment]
Danke für den Hinweis, ich habe gerade auf V2.0.2 geupdated und teste es mit dem Master-Slave-Zähler. Mir ist aber noch nicht ganz klar ob das tatsächlich das gewünschte Verhalten bringt. Ich würde das Ganze jetzt so konfigurieren:

- Eine Wärmepumpe mit Modbuszähler und aktiviertem Master. Master-Schalter und Slave-Schalter auf "muss eingeschaltet sein". Schalter ist "Immer eingeschaltet", damit alle Verbräuche ohne PV-Anhebung erfasst werden.
- Eine zusätzliche Wärmepumpe "PV-Anhebung" mit Slave-Zähler zufügen. Schalter wäre GPIO und hier stelle ich den Zeitplan für die PV-Anhebung ein.

Wäre das grundsätzlich das richtige Vorgehen? Wenn die WP z.B. schon aktuell mit 1KW läuft, würde der HomeManager diese Last nicht mit einbeziehen wenn es darum geht zu entscheiden, ob die PV-Anhebung geschaltet werden kann, oder? Wenn die PV-Anhebungs-WP auf 2,5KW Maximallast eingestellt ist, dann müssten trotzdem erst "freie" 2,5KW PV-Strom anliegen, damit dies geschaltet wird?

**Reply by camueller** [reply]
Damit das funktioniert, müßte die Konfiguration so aussehen:

- Eine Wärmepumpe mit Modbuszähler und aktiviertem Master. Master-Schalter nicht setzen, Slave-Schalter auf "muss eingeschaltet sein" (wie [Abbildung](https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/MasterSlaveMeter_DE.md)). Keinen Schalter konfigurieren, dann ist automatisch "Zählerbasierter Zustandsmelder" aktiv.

- Eine zusätzliche Wärmepumpe "PV-Anhebung" mit Slave-Zähler zufügen. Schalter wäre GPIO und hier stelle ich den Zeitplan für die PV-Anhebung ein.

Genau dieses Setup hatte ich im Sinn, als ich die Master-Slave-Zähler implementiert hatte. In den nächsten Tagen werde ich dieses Setup bei mir auch in Betrieb nehmen.

**Comment by realmuemmel** [comment]
Ich habe gerade auch die Master-Slave-Zähler getestet, allerdings habe ich das gleiche Problem wie in #205 nur mit Modbus. Sobald ich das Häkchen bei "Master-Zähler" setzte bekomme ich im Sunnyportal keine Verbräuche mehr übermittelt.  Sobald man das Häkchen entfernt funktioniert es wieder.

**Comment by camueller** [comment]
In der Master-Slave-Implementierung sind mindestens im Web-UI noch Fehler drin, wie ich selbst gerade bemerkt habe, nachdem ich jetzt auch meine zweiten produktiven SAE auf Version 2.0 aktualisiert habe. In den nächsten Tagen werde ich die Master-Slave-Implementierung selbst mit meiner WP testen und ggf. Fehler fixen für die nächste Version.

**Comment by camueller** [comment]
Auch im Backend der Master-Slave-Implementierung war ein Fehler, zumindest bei Verwendung von S0-Zählern. Dadurch wurden heute keine Verbräuche der WP erfasst, bis ich den Fehler bemerkt hatte und daraufhin erstmal Master-Slave wieder deaktiviert hatte.
Nachdem dieser Fehler gefixt und Master-Slave wieder aktviert ist, wird der Verbrauch via /semp korrekt gemeldet. Bereit für den nächsten realen Test morgen ...

**Reply by realmuemmel** [reply]
Ich habe gerade nochmal mit der 2.0.3 alles neu aufgesetzt und wie oben beschrieben ein Master-Slave-Setup mit WP und WP-PV-Anhebung konfiguriert. Allerdings melden die Zähler leider immernoch nichts zurück sobald man Master-Slave aktiviert. Dabei kommt im Log immer wieder die folgende Zeile (F-xxxxxxxx-000000000004-00 ist die Master-WP ohne PV-Anhebung):

2022-04-06 12:30:19,512 ERROR [Timer-0] d.a.s.m.ModbusSlave [ModbusSlave.java:92] F-xxxxxxxx-000000000004-00: Cannot connect to modbus CHA

Modbus ist aber korrekt konfiguriert und auch per "mbpoll -a 1 -r 342 -t 3:float -B -0 -1 localhost" abrufbar. Sobald ich Master-Slave deaktiviere funktioniert wieder alles einwandfrei, auch über den Modbus. Allerdings tauchen im Log dann diese Meldungen über alle Verbraucher auf (auch ohne Modbus);

2022-04-06 12:51:26,428 WARN [Timer-0] d.a.s.c.MeterReportingSwitch [MeterReportingSwitch.java:165] F-xxxxxxxx-000000000003-00: Meter not set.

Danke schonmal im vorraus, ich bin nach wie vor sehr glücklich dass es sowas wie SAE gibt :-)

**Reply by camueller** [reply]
Mit dem erfolgreichen Aufruf von "mbpoll ... localhost" hast Du nur überprüft, ob auf "localhost" ein Modbus/TCP-Prozess läuft (z.B. mbusd). Im SAE hast Du einen Modbus/TCP mit dem Namen "CHA" in den Einstellungen konfiguriert, bei dem ein anderer Hostname oder ein anderer Port angegeben ist als der, den mbpoll verwendet. Jedenfalls kann der SAE keine Verbindung dahin aufbauen. Das hat erstmal nichts mit Master-/Slave-Zähler zu tun, sondern ist ein einfach ein Konfigurationsfehler.

**Reply by realmuemmel** [reply]
Danke für die schnelle Antwort, aber es funktioniert ja alles sobald ich das Master-Slave Häckchen in SAE deaktiviere, auch der Modbus in SAE ohne Fehlermeldung ;-) Es werden korrekte Werte zum SHM übermittelt und angezeigt. Die Parameter für CHA in den SAE-Einstellungen und in mbusd sind auch exakt gleich, es wird auch der korrekte Zählerstand ausgegeben.

**Reply by realmuemmel** [reply]
Ich habe gerade testweise mal von Modbus auf S0 umgestellt, damit funktioniert alles einwandfrei, auch die Master-Slave umschaltung (per Raspberry pull-down). Top, das hilft mir schonmal!
Mit Modbus bekomme ich es damit aber leider nicht zum Laufen. Single klappt wie es soll, aber sobald ich das Häckchen bei Master-Slave setze kommt nichts mehr an und im Log kommt der Fehler "Cannot connect to modbus CHA".

**Reply by camueller** [reply]
Mit der 2.0.4 solle es auch mit Modbus klappen.
